abci/abci_test.go

```diff
@@ -768,3 +777,156 @@ func (suite *ABCITestSuite) isTopBidValid() bool {
+			MaxTxBytes: 1000000000000000000,
+		})
+		suite.Require().Len(res.Txs, 0)
+	})
+
+	suite.Run("accepts a single tx that is at the limit", func() {
+		suite.SetupTest() // reset
+		gasLimit := uint64(1000)
+		tx1, err := testutils.CreateTxWithLimit(suite.encodingConfig.TxConfig, suite.accounts[0], 0, 1000, testutils.CreateRandomMsgs(suite.accounts[0].Address, 1), gasLimit)
+		suite.Require().NoError(err)
+
+		mempool := mempool.NewAuctionMempool(suite.encodingConfig.TxConfig.TxDecoder(), suite.encodingConfig.TxConfig.TxEncoder(), 0, suite.config)
+		err = mempool.Insert(suite.ctx, tx1)
+		suite.Require().NoError(err)
+
+		suite.proposalHandler = abci.NewProposalHandler(
+			mempool,
+			log.NewTMLogger(os.Stdout),
+			suite.anteHandler,
+			suite.encodingConfig.TxConfig.TxEncoder(),
+			suite.encodingConfig.TxConfig.TxDecoder(),
+		)
+
+		// Update the consensus params to have a lower gas limit
+		cp := tmproto.ConsensusParams{
+			Block: &tmproto.BlockParams{
+				MaxGas: int64(1000),
+			},
+		}
+		ctx := suite.ctx.WithConsensusParams(&cp)
+
+		req := abcitypes.RequestPrepareProposal{
+			MaxTxBytes: 1000000000000000000,
+		}
+		res := suite.proposalHandler.PrepareProposalHandler()(ctx, req)
+
+		suite.Require().Len(res.Txs, 1)
+
+		decodedTx, err := suite.encodingConfig.TxConfig.TxEncoder()(tx1)
+		suite.Require().NoError(err)
+		suite.Require().Equal(res.Txs[0], decodedTx)
+	})
+
+	suite.Run("rejects a bid that is over the limit", func() {
+		suite.SetupTest() // reset
+		gasLimit := uint64(1000)
+		tx1, err := testutils.CreateAuctionTxWithSignersAndLimit(
+			suite.encodingConfig.TxConfig,
+			suite.accounts[0],
+			sdk.NewCoin("foo", sdk.NewInt(1000000000000000000)),
+			0,
+			1000,
+			suite.accounts[0:10],
+			gasLimit,
+		)
+		suite.Require().NoError(err)
+
+		mempool := mempool.NewAuctionMempool(suite.encodingConfig.TxConfig.TxDecoder(), suite.encodingConfig.TxConfig.TxEncoder(), 0, suite.config)
+		err = mempool.Insert(suite.ctx, tx1)
+		suite.Require().NoError(err)
+
+		suite.proposalHandler = abci.NewProposalHandler(
+			mempool,
+			log.NewTMLogger(os.Stdout),
+			suite.anteHandler,
+			suite.encodingConfig.TxConfig.TxEncoder(),
+			suite.encodingConfig.TxConfig.TxDecoder(),
+		)
+
+		// Update the consensus params to have a lower gas limit
+		cp := tmproto.ConsensusParams{
+			Block: &tmproto.BlockParams{
+				MaxGas: int64(1000) - 1,
+			},
+		}
+		ctx := suite.ctx.WithConsensusParams(&cp)
+
+		res := suite.proposalHandler.PrepareProposalHandler()(ctx, abcitypes.RequestPrepareProposal{
+			MaxTxBytes: 1000000000000000000,
+		})
+		suite.Require().Len(res.Txs, 0)
+	})
+
+	suite.Run("rejects a tx that is too large but accepts a smaller one", func() {
+		suite.SetupTest() // reset
+		gasLimit := uint64(1000)
+		tx1, err := testutils.CreateTxWithLimit(suite.encodingConfig.TxConfig, suite.accounts[0], 0, 1000, testutils.CreateRandomMsgs(suite.accounts[0].Address, 1), gasLimit)
+		suite.Require().NoError(err)
```

Comment:
If tx1 consumes `gasLimit` gas shldn't it have been selected (if it was ordered before tx2)?

Reply:
shld tx gas be gasLimit + 1? And how can we guarantee it's ordered before tx2?

Reply:
the gas limit is < tx1 gas limit while equal to gas limit of tx2.